test/SILOptimizer/capture_promotion_generic_context.sil

```diff
@@ -67,3 +67,51 @@ entry(%0 : $*T, %1 : $*U, %2 : $Int):
   %k = partial_apply %f<U>(%b) : $@convention(thin) <V> (@in V, <τ_0_0> { var τ_0_0 } <Int>) -> Int
   return %k : $@callee_owned (@in U) -> Int
 }
+
+enum E<X> {
+  case None
+  case Some(X)
+}
+
+struct R<T> {
+}
+
+// CHECK-LABEL: sil @_T023generic_promotable_box2Tf2nni_n : $@convention(thin) <T> (@in R<T>, @in Builtin.Int32, @owned E<(R<T>) -> Builtin.Int32>) -> () 
+// CHECK:       bb0(%0 : $*R<T>, %1 : $*Builtin.Int32, %2 : $E<(R<T>) -> Builtin.Int32>):
+// CHECK-NOT:     project_box
```

Comment:
I think a `CHECK:` line that ensures the promoted stack slot has type `$*E<(R<T>)->Int>` would be a good idea. LGTM otherwise.